File: src/main/java/ru/citeck/idea/CiteckPlugin.java
package ru.citeck.idea;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.Service;
import com.intellij.openapi.diagnostic.Logger;
import org.jetbrains.annotations.NotNull;

/**
 * Service required to allow use Disposer.register(CiteckPlugin.getInstance(), disposable)
 */
@Service(Service.Level.APP)
public final class CiteckPlugin implements Disposable {

    private static final Logger log = Logger.getInstance(CiteckPlugin.class);

    public static String ID = "ru.citeck.idea";

    public static @NotNull Disposable getInstance() {
        return ApplicationManager.getApplication().getService(CiteckPlugin.class);
    }

    @Override
    public void dispose() {
        log.info("Disposing CiteckPlugin");
    }
}

File: src/main/kotlin/ru/citeck/idea/CiteckPlugin.kt
package ru.citeck.idea

import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import ru.citeck.idea.artifacts.ArtifactsMetaLoader

/**
 * Service required to allow use Disposer.register(CiteckPlugin.getInstance(), disposable)
 */
@Service(Service.Level.APP)
class CiteckPlugin : Disposable {

    companion object {

        var ID: String = "ru.citeck.idea"

        private val log = Logger.getInstance(CiteckPlugin::class.java)

        @JvmStatic
        fun getInstance(): CiteckPlugin {
            return ApplicationManager.getApplication().getService(CiteckPlugin::class.java)
        }
    }

    override fun dispose() {
        log.info("Disposing CiteckPlugin")
    }

    class Initializer : ProjectActivity {
        override suspend fun execute(project: Project) {
            // Services initialized lazily. To force artifacts loading call getService.
            ApplicationManager.getApplication().getService(ArtifactsMetaLoader::class.java)
        }
    }
}
